lib/rewrite/rewrite-set-matches.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021 Balazs Scheidler <bazsi77@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "rewrite-set-matches.h"
+#include "scanner/list-scanner/list-scanner.h"
+#include "scratch-buffers.h"
+
+/* LogRewriteSetMatches
+ *
+ * This class implements the "set" expression in a rewrite rule.
+ */
+typedef struct _LogRewriteSetMatches LogRewriteSetMatches;
+
+struct _LogRewriteSetMatches
+{
+  LogRewrite super;
+  LogTemplate *value_template;
+  LogTemplateOptions template_options;
+};
+
+LogTemplateOptions *
+log_rewrite_set_matches_get_template_options(LogRewrite *s)
+{
+  LogRewriteSetMatches *self = (LogRewriteSetMatches *) s;
+  return &self->template_options;
+}
+
+static void
+log_rewrite_set_matches_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *path_options)
+{
+  LogRewriteSetMatches *self = (LogRewriteSetMatches *) s;
+  GString *result;
+  LogMessageValueType type;
+
+  result = scratch_buffers_alloc();
+
+  LogTemplateEvalOptions options = {&self->template_options, LTZ_SEND, 0, NULL, LM_VT_STRING};
+  log_template_format_value_and_type(self->value_template, *pmsg, &options, result, &type);
+
+  log_msg_make_writable(pmsg, path_options);
+
+  ListScanner scanner;
+
+  list_scanner_init(&scanner);
+  list_scanner_input_string(&scanner, result->str, result->len);
+  log_msg_clear_matches(*pmsg);
+  log_msg_unset_match(*pmsg, 0);
```

Comment:
This is just a question, but why do you explicitly unset `$0` here, while `unset-matches()` only does a `log_msg_clear_matches`?

`$*` doesn't care about `$0`, so I guess this is up to if we allow/expect that the individual matches (`$1`, `$2`, etc.) can be used separately.

I would use an actual unset than a simple clear, so the name-value pairs will be unset (empty).

Reply:
Good point.

I've removed this and added test coverage that log_msg_clear_matches() clears $0 too.

json-parser() and set-matches() sets the match variables starting from $1.. and $0 should be empty, but this is achieved by the log_msg_clear_matches() call (that sets num_matches to 0) and then eventually setting $1, which at
this point clears $0 lazily.

In the previous iteration $0 wouldn't have been unset and that's why it was present explicitly. But now with the explicit unsetting of matches, this is resolved.